packages/firestore/src/core/sync_engine.ts

```diff
@@ -770,29 +777,63 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
     const key = limboChange.key;
     if (!this.limboTargetsByKey.get(key)) {
       log.debug(LOG_TAG, 'New document in limbo: ' + key);
+      this.limboListenQueue.push(key);
+      this.pumpLimboResolutionListenQueue();
+    }
+  }
+
+  /**
+   * Starts listens for documents in limbo that are enqueued for resolution.
+   *
+   * When a document goes into limbo it is enqueued for resolution. This method
+   * repeatedly removes entries from the limbo resolution queue and starts a
+   * listen for them until either (1) the queue is empty, meaning that all
+   * documents that were in limbo either have active listens or have been
+   * resolved, or (2) the maximum number of concurrent limbo resolution listens
+   * has been reached.
+   *
+   * This method is invoked every time an entry is added to the limbo
+   * resolution queue and every time that a limbo resolution listen completes
+   * (either successfully or unsuccessfully). This ensures that all documents in
+   * limbo are eventually resolved.
+   *
+   * A maximum number of concurrent limbo resolution listens was implemented to
```

Comment:
This comment buries the lede: limiting concurrent limbo resolution is a top-line feature of this method. It also focuses on the change we made instead of focusing on documenting the intended behavior and what motivated that. In particular, it's not clear why "resource exhausted" errors are necessarily a problem. I suggest moving this into the first paragraph and emphasizing what this fixes. Something like this:

> Starts listens for documents in limbo that are enqueued for resolution, subject to a maximum number of concurrent resolutions. Without bounding the number of concurrent resolutions, the server can fail with resource exhausted errors which can lead to pathological client behavior as seen in https://github.com/firebase/firebase-js-sdk/issues/2683.